### PR TITLE
Skeleton UI

### LINKED ISSUE #19: UI skeleton
Add a skeleton UI to allow login and logout, and some non-functional payment UI

### PR DESCRIPTION
**Pull request by noisecapella**
#### What are the relevant tickets?
Fixes #19

#### What's this PR do?
- Adds backend logic to populate profile
- Adds frontend logic for mock UI. The payment button will submit an API request but no API exists currently.

#### How should this be manually tested?
Login, then type a number in the textbox. Open the network tab. Click 'Pay' and verify that the total you entered was sent to `/api/v0/payment/`.

### CODE REVIEW

File: static/js/lib/api_util.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
+// @flow
```
Comment (noisecapella) [inline static/js/lib/api_util.js]: None of this code is new, I just moved it around

File: static/js/lib/api_util_test.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
+/* global SETTINGS: false */
```
Comment (noisecapella) [inline static/js/lib/api_util_test.js]: These tests were only moved from `api_test.js`

**Comment by aliceriot** [comment]
functionality looks great! haha

File: static/js/actions/index.js [hunk 3]
```diff
@@ -1,6 +1,35 @@
...
+export const REQUEST_PAYMENT = 'REQUEST_PAYMENT';
+export const RECEIVE_PAYMENT_SUCCESS = 'RECEIVE_PAYMENT_SUCCESS';
+export const RECEIVE_PAYMENT_FAILURE = 'RECEIVE_PAYMENT_FAILURE';
+
+export const SET_TOTAL = 'SET_TOTAL';
+
+
+export const requestPayment = createAction(REQUEST_PAYMENT);
+export const receivePaymentSuccess = createAction(RECEIVE_PAYMENT_SUCCESS);
+export const receivePaymentFailure = createAction(RECEIVE_PAYMENT_FAILURE);
```
Comment (aliceriot) [inline static/js/actions/index.js]: nit: I like to group the action type declaration and the call to `createAction` together, like:

```js
export const ACTION_TYPE = 'ACTION_TYPE';
export const actionType = createAction(ACTION_TYPE);
```

File: static/js/containers/Payment.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    const {
+      ui: { total },
+      payment: { fetchStatus },
+    } = this.props;
+
+    const disabled = fetchStatus === FETCH_PROCESSING;
+
+    return <div className="payment">
+      <h3 className="intro">{SETTINGS.full_name}, Welcome to MIT Bootcamps</h3>
+      <h2 className="bootcamp-title">Internet of Things</h2>
```
Comment (aliceriot) [inline static/js/containers/Payment.js]: https://twitter.com/internetofshit/status/841932946504323072

**Comment by noisecapella** [comment]
@giocalitri I added a `Profile`. To try it out and have the name populated you'll need to delete all user objects first so they can be recreated

File: static/js/actions/index.js (hunk 3 above)
Comment (noisecapella) [inline static/js/actions/index.js]: Sure

File: static/js/reducers/index.js [hunk 5]
```diff
@@ -1,24 +1,52 @@
...
   default:
     return state;
   }
 };
 
 export default combineReducers({
-  checkbox
+  payment,
+  ui,
 });
```
Comment (noisecapella) [inline static/js/reducers/index.js]: I'm getting a strange error when installing redux-api:

    ➜  bootcamp-ecommerce git:(gs/payment) dwatch yarn add redux-api
    yarn add v0.19.1
    [1/4] Resolving packages...
    error An unexpected error occurred: "ENOENT: no such file or directory, open '/usr/local/lib/node_modules/yarn/node_modules/har-validator/lib/index.js'".
    info If you think this is a bug, please open a bug report with the information provided in "/src/yarn-error.log".
    info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
    ➜  bootcamp-ecommerce git:(gs/payment)

Comment (noisecapella) [inline static/js/reducers/index.js]: It seems like a quirk with yarn actually. I just installed it locally with a newer version of yarn and it seemed to install fine. I'll try it out

Comment (aliceriot) [inline static/js/reducers/index.js]: should we upgrade yarn? which version did you have trouble with?

Comment (noisecapella) [inline static/js/reducers/index.js]: 0.19.1. I'll make a separate PR to upgrade it

File: backends/utils.py [hunk 6]
```diff
@@ -2,12 +2,18 @@
 Utility functions for the backends
 """
 from datetime import datetime, timedelta
+import logging
 import pytz
 
+from django.core.exceptions import ObjectDoesNotExist
```
Comment (giocalitri) [inline backends/utils.py]: why the core exception?

File: bootcamp/settings.py [hunk 7]
```diff
@@ -114,7 +114,9 @@ def get_var(name, default):
     'raven.contrib.django.raven_compat',
 
     # Our INSTALLED_APPS
+    'backends',
```
Comment (giocalitri) [inline bootcamp/settings.py]: do we need to install the `backends` app?

File: static/js/reducers/index.js (hunk 5 above)
Comment (noisecapella) [inline static/js/reducers/index.js]: I tried out redux-api and it seems rough around the edges and redux-rest hasn't been updated in a year... do you know of any other options?

File: bootcamp/settings.py (hunk 7 above)
Comment (noisecapella) [inline bootcamp/settings.py]: I'm not sure but it's installed in micromasters so I figured it needed to be here too

File: backends/utils.py (hunk 6 above)
Comment (noisecapella) [inline backends/utils.py]: Copied from micromasters. I'll use a more specific one

**Comment by noisecapella** [comment]
@aliceriot I plagarized redux-api and made something similar. What do you think about this approach?

File: static/js/rest.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  _reducers[endpoint.name] = makeReducer(endpoint.name);
+
+  let fetchFunc = (...args) => {
+    let options = endpoint.makeOptions(...args);
+    return fetchJSONWithCSRF(endpoint.url, options);
+  };
+
+  _actions[endpoint.name] = (...params) => {
+    return dispatch => {
+      dispatch(createAction(makeRequestActionType(endpoint.name))());
```
Comment (aliceriot) [inline static/js/rest.js]: I think the created actions (the calls to `createAction` here and on lines 73 and 76) could probably be declared above and live in closure (avoiding having to re-define the little helper functions on every call)

Comment (aliceriot) [inline static/js/rest.js]: like

```js
let requestAction = createAction(makeRequestActionType(endpoint.name));
```

or something similar

Comment (noisecapella) [inline static/js/rest.js]: Sounds good. I think there will be other changes needed to deal with things like profiles which are stored by username but this should be a good starting point

Comment (aliceriot) [inline static/js/rest.js]: oh yeah. I wrote a little wrapper for `createAction` which handles some logic to automatically add in usernames to the actions, but if we wanted to abstract out things on the reducer end of things there would still be a little sorting out to do.

**Comment by noisecapella** [comment]
I made some more tweaks, let me know if you have any other comments there. Still need to fix flow

File: bootcamp/settings.py (hunk 7 above)
Comment (giocalitri) [inline bootcamp/settings.py]: ok

File: backends/utils.py (hunk 6 above)
Comment (giocalitri) [inline backends/utils.py]: maybe it's because I could not figure out a specific one there. feel free to leave as is

**Comment by giocalitri** [comment]
My part is fine as is

**Comment by giocalitri** [comment]
please expose the Profile in the Admin (maybe in another issue)

**Comment by giocalitri** [comment]
functionality looks good 
👍 for my part

**Comment by noisecapella** [comment]
Anything else @aliceriot?

**Comment by aliceriot** [comment]
sorry, meant to take a look and forgot!

**Pull request merged**